text/0000-wallet-replay-protection.md

```diff
@@ -0,0 +1,298 @@
+
+## Wallet rules for replay protection
+
+Most common transaction types are:
+1. A 1-2 `Regular` tx - 1 input, 2 outputs
+2. A 2-2 `PayJoin` tx - 2 inputs (one from each party), 2 outputs
+
+Let's call outputs that hang from a chain that leads to your anchor a `Protected` output. A transaction is protected from replays if a party contributes a Protected output as an input in the transaction. It's not enough that some other party contributes such an input. We are only really protected from replay attacks if we contribute the `Protected` output. Otherwise, our security depends on the honesty and following of the wallet rules of other participants which should not be relied on.
+
+Let's say that we want to be able to do both `Regular` and `PayJoin` transactions. We do both of these so we have some protected outputs and some that are not protected. The first problem we have is that after a wallet restore, we can't really tell whether an output is protected through a PayJoins+Anchor or not. It would be nice if we could label our outputs as either `Protected` or `Unprotected` after a wallet restore.
+
+### Separation of replay protected outputs
+
+Let's define a clear separation of outputs that are protected from those that are not. We define a protected output generator `GenP` that allows us to either `create` a protected output or `check` whether an output is `Protected`.
+
+An output is labeled as `Protected` *only if* it was created in a transaction where we contributed an input that was labeled as `Protected`. An exception to the rule are outputs that are created as a part of the _Bootstrap_ transaction discussed in the next section.
+
+#### Possible protected output generators implementations
+
+As we mentioned above, `GenP` for `Protected` outputs would need to have `create` and `check` functions defined.
+
+There are a few choices how to label outputs as `Protected` while still being able to identify them across difference devices:
+1. Call to `create` uses a new derivation path `P` that is used only for creating `Protected` outputs. Similarly `check` uses the same `P` to check whether an output is protected.
+2. Call to `create` creates a specific `r` value for `Protected` outputs e.g. they should start with `N` zeros or be divisible by some number `M`
+3. Call to `create` uses additional output information in the ~30 bytes that are available in the Bulletproofs to convey the idea whether the output is protected. Perhaps we could define a specific structure for these bytes e.g. `<scheme_version:1 byte><meta_data:4 bytes><data:25 bytes>` where `metadata` would also tell whether the `data` that follows is encrypted or not. The data could be encrypted using the `seed` key and could thus hold information on output labels which would only be available to the owner of the output. We could even include the starting bytes of the anchor that protects it if we wanted to. If we went such path, we would need to think of the possible drawbacks.

```

Comment:
How does that square with the 4 bytes: 0|type|switch_commitment_scheme|derivation_depth
from https://github.com/mimblewimble/grin-wallet/issues/105  ?

Reply:
I never dove into bulletproofs so I'm not sure yet. Might need to poke some people about this. Out of all 3 options, I prefer option 3. (labeling outputs via bulletproofs bytes) if it is possible because it means the information is directly on the output itself instead of label being some external property like a derivation path of the output. My current understanding from the linked issue is that we might have a "use in the future" byte available which is the first one called `0` in the `0|type|switch_commitment_scheme|derivation_depth`. I'll poke some people a bit as I'm not sure I'd be able to grasp this completely in a short time frame.